commit 9a98a283314581523327ae44442255d7eadb4900
Author: Jason Ryan <jasonwryan@gmail.com>
Date:   Thu Mar 5 10:32:32 2015 +1300

    Patched for new master

diff --git a/dwm.c b/dwm.c
index 8b67cea..9d3f527 100644
--- a/dwm.c
+++ b/dwm.c
@@ -1786,10 +1786,20 @@ toggletag(const Arg *arg) {
 void
 toggleview(const Arg *arg) {
 	unsigned int newtagset = selmon->tagset[selmon->seltags] ^ (arg->ui & TAGMASK);
+	Client *c = NULL;
+	unsigned int found = 0;
 
 	if(newtagset) {
 		selmon->tagset[selmon->seltags] = newtagset;
-		focus(NULL);
+
+		for(c = selmon->clients; c && !(found = c->tags & arg->ui); c = c->next);
+		if(found && ISVISIBLE(c)) {
+			detach(c);
+			attach(c);
+			focus(c);
+		}
+		else
+			focus(NULL);
 		arrange(selmon);
 	}
 }
